doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
+platforms (ones that support setting arbitrary interrupt timeouts) are
+expected to be in the range of 10 kHz, with software emulation
+platforms and legacy drivers using a more traditional 100 Hz value.
+
+Conversion
+----------
+
+Zephyr provides an extensively enumerated conversion library with
+rounding control for all time units.  Any unit of "ms" (milliseconds),
+"us" (microseconds), "tick", or "cyc" can be converted to any other.
+Control of rounding is provided, and each conversion is available in
+"floor" (round down to nearest output unit), "ceil" (round up) and
+"near" (round to nearest).  Finally the output precision can be
+specified as either 32 or 64 bits.
+
+For example: ``k_ms_to_ticks_ceil32()`` will convert a millisecond
+input value to the next higher number of ticks, returning a result
+truncated to 32 bits of precision; and ``k_cyc_to_us_floor64()`` will
+convert a measured cycle count to an elapsed number of microseconds in
+a full 64 bits of precision.  See the reference documentation for the
+full enumeration of conversion routines.
+
+On most platforms, where the various counter rates are integral
+multiples of each other and where the output fits within a single
+word, these conversions expand to a 2-4 operation sequence, requiring
+full precision only where actually required and requested.
+
+Uptime
+======
+
+The kernel tracks a system uptime count on behalf of the application.
+This is available at all times via ``k_uptime_get()``, which provides
+an uptime value in milliseconds since system boot.  This is expected
+to be the utility used by most portable application code.
+
+The internal tracking, however, is as a 64 bit integer count of ticks.
+Apps with precise timing requirements (that are willing to do their
+own conversions to portable real time units) may access this with
+``k_uptime_ticks()``.
+
+Timeouts
+========
+
+The Zephyr kernel provides many APIs with a "timeout" parameter.
+Conceptually, this indicates the time at which an event will occur.
+For example:
+
+* Kernel blocking operations like ``k_sem_take()`` or
+  ``k_queue_get()`` may provide a timeout after which the routine will
+  return with an error code if no data is available.
+
+* Kernel ``struct k_timer`` objects must specify delays for their
+  duration and period.
+
+* The kernel ``k_delayed_work`` API provides a timeout parameter
+  indicating when a work queue item will be added to the system queue.
+
+All these values are specified using a ``k_timeout_t`` value.  This is
+an opaque struct type that must be initialized using one of a family
+of ``K_TIMEOUT`` macros.  The most common, ``K_TIMEOUT_MS()``, defines
+a time in milliseconds after the current time (strictly: the time at
+which the kernel receives the timeout value).
+
+Other options for timeout initialization follow the unit conventions
+described above: ``K_TIMEOUT_US()``, ``K_TIMEOUT_TICKS()`` and
+``K_TIMEOUT_CYC()`` specify timeout values that will expire after
```

Comment:
Not really related to this PR but just pondering why we have K_TIMEOUT_CYC() instead of K_TIMEOUT_CYCLE(). Saving two bytes here seems unnecessary as we have K_TIMEOUT_TICKS() which is identical length.